### PR TITLE
Tune GC heuristics for big remsets.

### PR DESCRIPTION
**Pull request by jiahao**
Take the number of pointers transgressing the generational frontier
into account when deciding wether to do a young or an old gen
collection. Helps #10428. Didn't run the bench suite yet.

### CODE REVIEW

**Comment by yuyichao** [comment]
Most of the conflicts seems to be cause by my reorganization for gc debugging. Will push a rebased version later.

**Comment by yuyichao** [comment]
The original branch is backed up at [the yyc/gctune.bak branch](https://github.com/JuliaLang/julia/tree/yyc/gctune.bak) with the commit SHA 238ed08b0c4f45dc7afec0f3ece2b6c3f37330c2.

Rebased version force pushed to the ob/gctune branch.

**Comment by yuyichao** [comment]
@carnaval

File: src/gc.c [hunk 1]
```diff
@@ -1708,10 +1715,12 @@ static int push_root(jl_value_t *v, int d, int bits)
         }
     }
     else if (vt == (jl_value_t*)jl_module_type) {
+        // should increase nptr here
```
Comment (vtjnash) [inline src/gc.c]: is this an action item, or a comment on what it does, or neither?

**Comment by JeffBezanson** [comment]
Cool! How much does it help readdlm?

**Comment by carnaval** [comment]
so this is the change that has been rotting away on one of my branches for a long time. Last time I tried it solved the runaway heuristic thing, and brought down the time from 95% gc to around 20% in my recollection.

I probably wont have time to test it again until next week.

**Comment by jakebolewski** [comment]
Bump.  It would be nice to run the performance tests and get this in for 0.4.

**Comment by simonster** [comment]
i take full responsability

**Comment by simonster** [comment]
@carnaval is drunk

File: src/gc.c (hunk 1 above)
Comment (ViralBShah) [inline src/gc.c]: This does look like a dangerous comment.

**Pull request merged**